### PR TITLE
Switch to a date-based catalog versioning system, and related updates

### LINKED ISSUE #191: Default to a "latest" version?
The ACCESS-NRI Intake catalog is versioned in the same way as the `access-nri-intake` package (see `/g/data/xp65/public/apps/access-nri-intake-catalog`).

The default version is equal to the version of `access-nri-intake` being used, though the version can be specified when the catalog is loaded, e.g.

```python
import intake

cat = intake.cat.access_nri
cat.metadata["version"]

cat_another_version = intake.cat.access_nri(version="v0.1.0")
cat_another_version.metadata["version"]
```

This default behaviour ensures that the catalog version and the version of `access-nri-intake` being used are compatible. However, it also means that users with old versions of `access-nri-intake` installed won't use the most recent version of the catalog by default.

We should consider instead defaulting to a "latest" version of the catalog that symlinks to the most recent released version. To manage compatibility issues, we could have "latest" versions for each major release (e.g. `latest_v1`) and ensure we increment the major version when changes are introduced to `access-nri-intake` that break compatibility with old catalogs.

### PR DESCRIPTION
**Pull request by marc-white**
Closes #191 .

This PR significantly alters the way the catalog generated by `access-nri-intake-catalog` is handled, stored, and versioned.

### Catalog storage location

The live catalog will now be stored on `gdata/xp65` (`access_nri_intake.CATALOG_LOCATION`), rather than being shipped with the package. This is because all of the catalog data lives on Gadi anyway, so there doesn't seem to be much point to being able to see the catalog definition YAML, but not be able to access any data.

However, to support power users and local developers, there is the option to place a `catalog.yaml` file at a defined location in the user's home directory (`access_nri_intake.USER_CATALOG_LOCATION`). This catalog will automatically take precedence over the 'live' catalog on Gadi, if it exists. I thought about adding utility functions that would allow users to create this directory, put a `catalog.yaml` in it, etc.; however, given that it's a power user move, I decided it's easier and safer to have the user do that themselves manually.

### Versioning

Catalog versions will now be date-based, e.g., a catalog built today will be, by default, `v2024-11-07`. Attempts to set a version number that doesn't conform to this pattern will raise an exception.

`catalog.yaml` will now contain a `min` and `max` version. This is to cover the possibility that the catalog structure may change, and a particular version of `catalog.yaml` may not be compatible with certain catalog versions. I've confirmed that doing the usual `<` and `>` operations on our version strings has the expected output.

Under this versioning schema, there isn't really a need to have a symlinked `latest` version of the catalog (and also because we can update the live catalog on `xp65` at will, without doing a code release).

### Building the catalog

Because `catalog.yaml` will be placed on `xp65` now, there is no need to make a new code release for every catalog build.

As mentioned above, a catalog built today will take today's date as the default version, although the user can override it.

The build process is now intelligent to the presence of older versions of `catalog.yaml`, and to directories that look like older catalogs:
- If no `catalog.yaml` exists, one will be created. If the data directory contains folders that look like catalog versions (i.e. vYYYY-MM-DD), then the code will use those to construct the `min` and `max` version boundaries (i.e., it will assume that the new `catalog.yaml` is good for describing those existing sources).
- If a `catalog.yaml` exists, and there is no structural change to the catalog, then the existing `catalog.yaml` will be updated with:
  - A new default (and probably max) version; and,
  - Updated storage metadata that is the union of the old and new storage metadata (i.e., we deliberately don't remove any storage metadata that is redundant with the new catalog version, as it will still be required for the older versions under this `catalog.yaml`).
- If the new catalog is structurally different to the old catalog, then a new `catalog.yaml` will be created, with min version = max version = current/new version. The old `catalog.yaml` will be moved aside to `catalog-<old min version>-<old max version>.yaml`. These catalogs are nominally not accessible, unless the user hacks the `access_nri_intake.CATALOG_LOCATION` variable. For now, given how infrequently we'll be making the sort of changes that will trigger this scenario, I'm fine with that.

### Documentation

To be updated once we're happy with the core structure of this update. The updates required will be:
- Alter the catalog build process instructions;
- Document the ability to have a local 'override' `catalog.yaml`.

### Testing

All of the above should have at least one unit test.

### CODE REVIEW

**Comment by rbeucher** [comment]
OK. What's the best way for us to test this on GADI?

**Comment by marc-white** [comment]
> OK. What's the best way for us to test this on GADI?

The way I've done this in the past (e.g. to test #175 updates):

1. Check out the code onto Gadi, and do a local install of the package (non-destructive override of the environment version)
2. Copy the `build_all.sh` script, and modify it to point to a dummy directory, a subset of the data sources, etc.
3. Run the copied build process as per the documentation
4. Scream incoherently when it fails (optional), work out why it fails, run it again 
5. Repeat 3-4

The main thing to check would be to copy the existing catalog folders (e.g. v0.1.3) to the dummt directory, create symlinks to them using the date catalog format, and make sure the build process picks those up as previous versions and sets the catalog `min` version correctly.

**Comment by rbeucher** [comment]
I have just tried with only one source in `cmip6.yml` same pb

**Comment by marc-white** [comment]
Do you have an existing `catalog.yaml` in the build directory? If so, it's going to look for a `min` value in that, and it appears to be bombing out because it can't find one. The assumption I made is that the first catalog build in the 'live' directory on `xp65` (and, therefore, the first build in your dummy area) will be the first `catalog.yaml` to appear there, so it will inspect the present directories to compute a `min` and `max` version instead.

**Comment by rbeucher** [comment]
I have tried with an empty folder and I still get the issue...

**Comment by marc-white** [comment]
OK I'll take a look and see what's happening...

**Comment by marc-white** [comment]
Found the problem - the code is building the catalog in one place, but it's looking for the existing catalog (and, it seems, building the catalog.yaml back out to) the live location referred to by the `CATALOG_LOCATION` variable. I need to have a quick think about how to rectify that.

**Comment by marc-white** [comment]
@rbeucher OK, check out the latest code version and try again. You'll need to update your test version of the `build_all.sh` script to include the new `--catalog_base_path` option (see the canonical `build_all.sh` in the repo).

The issue was that, while changing the `build_base_path` option in the build script lets you stick the catalog _data_ in a different place, the catalog _yaml_ was always being forced into either live location on `xp65`, or into your home area if you have a local `catalog.yaml` set up. I've now added an extra option to the `build()` function call, as well as some additional supporting changes, that let you set where the catalog.yaml lives. In production, this will be the same place as the catalog data, hence why the canonical `build_all.sh` script sets `--build_base_path` and `--catalog_base_path` to be the same.

File: tests/test_cli.py
Comment (charles-turner-1) [inline tests/test_cli.py:331]: As above about looping

File: tests/test_cli.py
Comment (charles-turner-1) [inline tests/test_cli.py:403]: As above about looping

File: tests/test_data.py
Comment (charles-turner-1) [inline tests/test_data.py:86]: I wonder if we might want to use strptime instead of a regex to parse the version in order to avoid letting through these slightly wrong dates? Probably unimportant

**Review by charles-turner-1 — COMMENTED** [review]
Looks good to me - I haven't run a build test on Gadi yet so I'll hold off approval.

Left a few comments but they're mostly minor nitpicks - looks like a solid implementation all in all.

File: src/access_nri_intake/cli.py
Comment (marc-white) [reply src/access_nri_intake/cli.py:323]: Yeah, I wondered about that. The 'completely new catalog with no existing data sources' case in the tests should cover that, so I'm hoping it's just a codecov gremlin.

File: tests/test_data.py
Comment (marc-white) [reply tests/test_data.py:86]: If we really cared about having 'real' dates, we'd need a two-part validation - one to check that the version string starts with `v`, and then a second to check the date, possibly with `strptime`. I'm tempted to not bother given that the date really represents a version number, so it doesn't have to be a 'real' date, but I'm then tempted to make this change to avoid any confusion.

File: src/access_nri_intake/data/__init__.py
Comment (marc-white) [reply src/access_nri_intake/data/__init__.py:20]: Good idea, didn't know that was an option. Updated.

File: src/access_nri_intake/cli.py
Comment (marc-white) [reply src/access_nri_intake/cli.py:334]: Probably a Friday afternoon brain issue, but how would returning `d` instead of doing in-place modifications give us an alert of the schema changing?

Comment (charles-turner-1) [reply src/access_nri_intake/cli.py:334]: Sorry, I could have been clearer - by returning d, we could write a test that explicitly checks the returned dictionary (solves the codecov gremlin above) - which would also presumably fail due to a KeyError if the schema were to change.

Comment (marc-white) [reply src/access_nri_intake/cli.py:334]: Ah yeah, I see it now. OK, that seems like a reasonable idea, will implement.

File: tests/test_data.py
Comment (charles-turner-1) [reply tests/test_data.py:86]: The datetime module will let you stick prefixes on date strings, eg:

```python
>>> v_str = "v2024-10-13"
>>> datetime.strptime(v_str,"v%Y-%m-%d")
 datetime.datetime(2024, 10, 13, 0, 0)
>>> no_v_str = "2024-10-13"
>>> datetime.strptime(no_v_str ,"v%Y-%m-%d")
ValueError: time data '2024-10-13' does not match format 'v%Y-%m-%d'
```

But yeah, I'm inclined to agree that its unimportant - I wouldn't have even thought about it if the test wasn't there.

File: src/access_nri_intake/cli.py
Comment (marc-white) [reply src/access_nri_intake/cli.py:323]: Seems to have been fixed by the update below.

File: tests/test_cli.py
Comment (marc-white) [reply tests/test_cli.py:276]: The looping does make sense; just means the logic for determining what should be the min and max values would be moved out to the line defining the for loop.

**Comment by charles-turner-1** [comment]
Running a build test right now - will comment if I have any issues

**Comment by charles-turner-1** [comment]
@rbeucher It looks like you might have overwritten a default catalog location or similar on Friday? I'm currently getting the following error - are you able to take a look & let me know if the file in the error message is one you recognise?

```python
>>> import intake
>>> intake.cat.access_nri
/g/data/hh5/public/apps/miniconda3/envs/analysis3-24.07/lib/python3.10/importlib/__init__.py:126](https://are.nci.org.au/g/data/hh5/public/apps/miniconda3/envs/analysis3-24.07/lib/python3.10/importlib/__init__.py#line=125): RuntimeWarning: Unable to access a default catalog location. Calling intake.cat.access_nri will not work.
  return _bootstrap._gcd_import(name[level:], package, level)
access_nri:
  args: {}
  description: ''
  driver: intake.catalog.base.Catalog
  metadata: {}

>>> from access_nri_intake.utils import get_catalog_fp
>>> get_catalog_fp()
'/g/data/xp65/public/apps/access-nri-intake-catalog/catalog.yaml'
>>> intake.open_catalog('/g/data/xp65/public/apps/access-nri-intake-catalog/catalog.yaml').access_nri
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
Cell In[9], line 1
----> 1 intake.open_catalog('[/g/data/xp65/public/apps/access-nri-intake-catalog/catalog.yaml](https://are.nci.org.au/g/data/xp65/public/apps/access-nri-intake-catalog/catalog.yaml)').access_nri
...
FileNotFoundError: [Errno 2] No such file or directory: '[/g/data/tm70/rb5533/intake_tests/v0.1.4](https://are.nci.org.au/g/data/tm70/rb5533/intake_tests/v0.1.4)+13.g869f576.dirty[/metacatalog.csv](https://are.nci.org.au/metacatalog.csv)'
```

I've run a catalog build with just CMIP5 enabled which built without any issues - I'm just running into this slightly weird error trying to import it.

**Comment by rbeucher** [comment]
Yes. Sorry, my bad. It looks like I have mess up with the file

**Comment by charles-turner-1** [comment]
No worries. I'm looking into it more closely now - probably something we're going to want to have some guard rails against.

**Comment by charles-turner-1** [comment]
I've restored the file to default to v0.1.3 for now.

**Comment by marc-white** [comment]
@charles-turner-1 I think you might have to hack the code to open your existing catalog (or, better yet, put it into `~/.access_nri_intake_catalog/`) - the code only knows to look for a catalog to open at either `CATALOG_LOCATION = "/g/data/xp65/public/apps/access-nri-intake-catalog/catalog.yaml"` or `USER_CATALOG_LOCATION = Path.home() / ".access_nri_intake_catalog/catalog.yaml"`.

We should probably look at adding a `path` option to `intake.cat.access_nri` to see if we can make that more flexible.

**Review by charles-turner-1 — APPROVED** [review]
I've managed to build a catalog subset & I'm happy that this behaves as expected. 

I think it might be nice to add a toggle that allows the user to toggle discovery of `$HOME/.access_nri_intake_catalog/catalog.yaml` from the python interpreter - rather than requiring the user to rename the file/folder to disable it overriding the default, but perhaps that's for a separate PR?

**Comment by charles-turner-1** [comment]
Sorry @marc-white, just saw your comment - the default catalog was pointing at a nonexistent file. My catalog in `.access_nri_intake_catalog/catalog.yaml` worked just fine.

I've approved but I think before merging we want to update docs?

**Comment by marc-white** [comment]
Yes, a docs update is definitely required. I can get started on that today.

**Comment by marc-white** [comment]
I've added the necessary documentation, take a peek...

File: docs/management/schema.rst
Comment (charles-turner-1) [inline docs/management/schema.rst:6]: I think this should be version 0.1.4, not 0.14

File: docs/management/building.rst
Comment (marc-white) [reply docs/management/building.rst:151]: Correct

File: docs/management/schema.rst
Comment (marc-white) [reply docs/management/schema.rst:6]: Correct, I think there might be a couple of those...

**Comment by marc-white** [comment]
> Other than the two minor typos, documentation changes look correct to me.
> 
> I updated `cli.py::build` `if update: ...` clause on my local machine to make it a bit more readable to check the docs - are you happy for me to push the changes to the branch?

Yep push the changes

**Comment by charles-turner-1** [comment]
@marc-white Can you confirm the shorthand variable names I used in [5c80be3](https://github.com/ACCESS-NRI/access-nri-intake-catalog/pull/243/commits/5c80be31498d003a0ea8f80a018985852d65f8f1) aren't misleading? Other than that I think this is good to go.

**Comment by marc-white** [comment]
> @marc-white Can you confirm the shorthand variable names I used in [5c80be3](https://github.com/ACCESS-NRI/access-nri-intake-catalog/pull/243/commits/5c80be31498d003a0ea8f80a018985852d65f8f1) aren't misleading? Specifically `driver_new`, etc..

Looks reasonable to me!

**Comment by marc-white** [comment]
I'm going to get onto Gadi now and create symlinks for the existing catalog versions, so they get picked up the first time we generate a 'new-style' catalog.

**Pull request merged**
